lib/fs/basicfs_copy_range_ioctl.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// +build !windows,!solaris,!darwin
+
+package fs
+
+import (
+	"syscall"
+	"unsafe"
+)
+
+func init() {
+	registerCopyRangeImplementation(copyRangeImplementation{
+		name: "ioctl",
+		impl: copyRangeIoctl,
+	})
+}
+
+const FICLONERANGE = 0x4020940d
+
+/*
+http://man7.org/linux/man-pages/man2/ioctl_ficlonerange.2.html
+
+struct file_clone_range {
+   __s64 src_fd;
+   __u64 src_offset;
+   __u64 src_length;
+   __u64 dest_offset;
+};
+*/
+type fileCloneRange struct {
+	srcFd     int64
+	srcOffset uint64
+	srcLength uint64
+	dstOffset uint64
+}
+
+func copyRangeIoctl(src, dst basicFile, srcOffset, dstOffset, size int64) error {
```

Comment:
One thing to note (somewhere) about this one is that the write must be disk block aligned in both offset and size. Luckily, we generally deal with blocks that are larger than the disk block size so this is mostly fine. But it means we can't use this to copy an entire file, unless that file happens to be an even number of blocks long. For example, this fails pretty much all the tests:

```
diff --git a/lib/fs/basicfs_copy_range_test.go b/lib/fs/basicfs_copy_range_test.go
index a75c2655..f23d6390 100644
--- a/lib/fs/basicfs_copy_range_test.go
+++ b/lib/fs/basicfs_copy_range_test.go
@@ -18,7 +18,7 @@ import (
 )

 var (
-       generationSize  int64 = 4 << 20
+       generationSize  int64 = 4 << 20 + 1
        defaultCopySize int64 = 1 << 20
```